### ISSUE TITLE
locally modifying+testing published definitions is a pain (bug/rant/perhaps-question)

### ISSUE THREAD

**Issue by jasonswearingen** [Microsoft/TypeScript#14321]
the ```@types``` process is pretty great for consuming .d.ts files (much better than the previous typings or tsd workflows)   

however, I seem to have a big problem as a creator of ```.d.ts``` for publication.    I don't see any way of creating a .d.ts file for local use and testing, and then be able to then publish the file as-is.   

For example, if this is a definition I publish to DT:  

```foo/index.d.ts```:
`
declare namespace Foo{ export function foo():void; }
export = Foo;
export as namespace Foo;
`

and usage:
`
import foo = require("foo");
`

How would I consume my own definitions for ```foo``` locally?  I don't see how to do this with the normal import style.  instead I need to modify my d.ts and consumption usage like:

```myProject/_dts/foo.ts```
`
export declare namespace _foo{ export function foo():void; }
export const Foo: typeof _foo = require("foo");
`
and usage:
`
import foo = require("./_dts/foo");
`

Basically this means that I can't easily test and/or modify a definition that I want to share publically, which adds a lot of friction for re-contributing to the community.   Am I doing something wrong?  Is there an easy way to modify and locally reference a published type?

**Note**: I don't know if i'm doing it wrong these days, so please correct my understanding if you see problems with my logic above.

**Comment by jasonswearingen** [comment]
by the way, pre ```@types``` I published perhaps 10 ```d.ts``` on DefinitelyTyped, but based on the issues described in my above rant the friction to publish+maintain typings is too great, so I just write definitions for private use.   

basically from a consumption standpoint, ```@types``` is great, but from a producer's perspective it is worse than the previous DT ```tsd``` model.

**Comment by mhegazy** [comment]
> How would I consume my own definitions for foo locally? I don't see how to do this with the normal import style. instead I need to modify my d.ts and consumption usage like:

Let's say you want to write defintions for `foo`, which is a module:
- under your app: `src\extra-types\foo`, create `index.d.ts` that has your declaration as noted above (along with tsconfig.json, tslint.json, and tests for DT if you want). 
- in your `tsconfig.json` add a path mapping rule: 
 ```js
{
     ...
    "paths": {
       "foo" : ["src/extra-types/foo"]
    }
}
  ```
  or if you have multiple:
  ```js
{
     ...
    "paths": {
       "*" : [ "*", "src/extra-types/*"]
    }
}
  ```
- once you have published to DT, delete `src\extra-types\foo` and add `@types/foo`, no other changes should be required.

**Comment by jasonswearingen** [comment]
ok thank you for those details.  I finally got around to writing a new definition and see that your info works.

are those in the docs somewhere?
